crates/matrix-sdk/src/event_cache/room/mod.rs

```diff
@@ -409,6 +399,22 @@ impl RoomEventCacheInner {
             all_duplicates,
         ) = state.collect_valid_and_duplicated_events(timeline.events).await?;
 
+        // If the timeline isn't limited, and we already knew about some past events,
+        // then this definitely know what the timeline head is (either we know
+        // about all the events persisted in storage, or we have a gap
+        // somewhere). In this case, we can ditch the previous-batch
+        // token, which is an optimization to avoid unnecessary future back-pagination
+        // requests.
+        //
+        // We can also ditch it, if we knew about all the events that came from sync,
+        // viz. they were all deduplicated. In this case, using the
+        // previous-batch token would only result in fetching other events we
+        // knew about. This is slightly incorrect in the presence of
+        // network splits, but this has shown to be Good Enough™.
+        if !timeline.limited && state.events().events().next().is_some() || all_duplicates {
```

Comment:
We should do something about these `events.events()`. I see it returns an iterator so perhaps we should call it just that `events.iterator()`.